File: app/src/main/java/cl/figonzal/evaluatool/evalua4/Evalua4Activity.java
/*-----------------------------------------------------------------------------
                                                                              -
 This file is subject to the terms and conditions defined in                  -
 file 'LICENSE', which is part of this source code package                    -
                                                                              -
 Autor: Felipe González                                                       -
 Email: felipe.gonzalezalarcon94@gmail.com                                    -
                                                                              -
 Copyright (c) 2020                                                           -
                                                                              -
 Last modified 27-07-20 23:14                                                 -
 -----------------------------------------------------------------------------*/

package cl.figonzal.evaluatool.evalua4;

import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.crashlytics.FirebaseCrashlytics;

import java.util.ArrayList;
import java.util.List;

import cl.figonzal.evaluatool.ConfigRoutes;
import cl.figonzal.evaluatool.R;
import cl.figonzal.evaluatool.Utilidades;
import cl.figonzal.evaluatool.adapter.EvaluaAdapter;
import cl.figonzal.evaluatool.modelo.Evalua;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionAdapter;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionedRecyclerViewAdapter;

public class Evalua4Activity extends AppCompatActivity implements EvaluaAdapter.ClickListener {

    private SectionedRecyclerViewAdapter sectionedRecyclerViewAdapter;

    private FirebaseCrashlytics crashlytics;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_evalua4);
        crashlytics = FirebaseCrashlytics.getInstance();

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitleTextColor(getResources().getColor(R.color.colorOnPrimary, getTheme()));
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        assert actionBar != null;
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_white_24dp);
        actionBar.setTitle(getString(R.string.TOOLBAR_EVALUA_4));

        configurarExpandedList();
    }

    private void configurarExpandedList() {
        sectionedRecyclerViewAdapter = new SectionedRecyclerViewAdapter();

        //SUBTIMES - SUBMODULO 1
        List<Evalua> subItems1 = new ArrayList<>();
        subItems1.add(new Evalua(getString(R.string.EVALUA_4_M1_SI_1)));

        List<Evalua> subItems2 = new ArrayList<>();
        subItems2.add(new Evalua(getString(R.string.EVALUA_4_M2_SI_1)));
        subItems2.add(new Evalua(getString(R.string.EVALUA_4_M2_SI_2)));
        subItems2.add(new Evalua(getString(R.string.EVALUA_4_M2_SI_3)));
        subItems2.add(new Evalua(getString(R.string.EVALUA_4_EVALUA_GLOBAL)));
        subItems2.add(new Evalua(getString(R.string.EVALUA_4_M2_SI_4)));

        List<Evalua> subItems3 = new ArrayList<>();
        subItems3.add(new Evalua(getString(R.string.EVALUA_4_M3_SI_1)));

        List<Evalua> subItems4 = new ArrayList<>();
        subItems4.add(new Evalua(getString(R.string.EVALUA_4_M4_SI_1)));
        subItems4.add(new Evalua(getString(R.string.EVALUA_4_M4_SI_2)));
        subItems4.add(new Evalua(getString(R.string.EVALUA_4_EVALUA_GLOBAL)));
        subItems4.add(new Evalua(getString(R.string.EVALUA_4_M4_SI_3)));

        List<Evalua> subItems5 = new ArrayList<>();
        subItems5.add(new Evalua(getString(R.string.EVALUA_4_M5_SI_1)));

        List<Evalua> subItems6 = new ArrayList<>();
        subItems6.add(new Evalua(getString(R.string.EVALUA_4_M6_SI_1)));
        subItems6.add(new Evalua(getString(R.string.EVALUA_4_M6_SI_2)));
        subItems6.add(new Evalua(getString(R.string.EVALUA_4_EVALUA_GLOBAL)));
        subItems6.add(new Evalua(getString(R.string.EVALUA_4_M6_SI_4)));

        //HEADERS
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_4_MODULO_1), subItems1, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_4_MODULO_2), subItems2, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_4_MODULO_3), subItems3, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_4_MODULO_4), subItems4, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_4_MODULO_5), subItems5, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_4_MODULO_6), subItems6, this));

        final RecyclerView recyclerView = findViewById(R.id.recyclerview);
        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        recyclerView.setAdapter(sectionedRecyclerViewAdapter);
    }

    @Override
    public void onItemRootViewClicked(@NonNull String sectionTitle, int itemAdapterPosition) {

        Utilidades.handleRoutes(
                new ConfigRoutes(this).getRouteMapEvalua4(),
                sectionTitle,
                sectionedRecyclerViewAdapter,
                itemAdapterPosition,
                this
        );
    }

    @Override
    public void onHeaderRootViewClicked(@NonNull final EvaluaAdapter section) {

        SectionAdapter sectionAdapter = sectionedRecyclerViewAdapter.getAdapterForSection(section);

        final boolean wasExpanded = section.isExpanded();
        final int previousItemsTotal = section.getContentItemsTotal();

        section.setExpanded(!wasExpanded);
        sectionAdapter.notifyHeaderChanged();

        if (wasExpanded) {
            sectionAdapter.notifyItemRangeRemoved(0, previousItemsTotal);
        } else {
            sectionAdapter.notifyAllItemsInserted();
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {

        if (item.getItemId() == android.R.id.home) {

            Log.d(getString(R.string.CLOSE_EVALUA_4), getString(R.string.ACTIVIDAD_CERRADA));

            crashlytics.log(getString(R.string.CLOSE_EVALUA_4) + getString(R.string.ACTIVIDAD_CERRADA));

            finish();

            return true;
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/cl/figonzal/evaluatool/evalua/evalua4/Evalua4Activity.kt
/*

 This file is subject to the terms and conditions defined in
 file 'LICENSE', which is part of this source code package

 Autor: Felipe González
 Email: felipe.gonzalezalarcon94@gmail.com

 Copyright (c) 2021

 Last modified 16-08-21 16:35
 */
package cl.figonzal.evaluatool.evalua.evalua4

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import cl.figonzal.evaluatool.R
import cl.figonzal.evaluatool.adapter.HeaderAdapter
import cl.figonzal.evaluatool.databinding.ActivityEvalua4Binding
import cl.figonzal.evaluatool.model.Child
import cl.figonzal.evaluatool.model.Header
import cl.figonzal.evaluatool.utilities.configureActionBar
import cl.figonzal.evaluatool.utilities.configureFabWsp
import timber.log.Timber
import java.util.*

class Evalua4Activity : AppCompatActivity() {

    private lateinit var binding: ActivityEvalua4Binding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityEvalua4Binding.inflate(layoutInflater)
        setContentView(binding.root)

        configureActionBar(R.string.TOOLBAR_EVALUA_4, binding.includeToolbar.materialToolbar)
        configureFabWsp(binding.fabWsp)

        val headers = listOf(
            Header(getString(R.string.EVALUA_4_MODULO_1)),
            Header(getString(R.string.EVALUA_4_MODULO_2)),
            Header(getString(R.string.EVALUA_4_MODULO_3)),
            Header(getString(R.string.EVALUA_4_MODULO_4)),
            Header(getString(R.string.EVALUA_4_MODULO_5)),
            Header(getString(R.string.EVALUA_4_MODULO_6)),
        )

        val subItemsList = listOf(
            listOf(
                Child(getString(R.string.EVALUA_4_M1_SI_1))
            ),
            listOf(
                Child(getString(R.string.EVALUA_4_M2_SI_1)),
                Child(getString(R.string.EVALUA_4_M2_SI_2)),
                Child(getString(R.string.EVALUA_4_M2_SI_3)),
                Child(getString(R.string.EVALUA_4_VALORACION_GLOBAL_RAZONAMIENTO)),
                Child(getString(R.string.EVALUA_4_INDICE_GENERAL_COGNITIVO))
            ),
            listOf(
                Child(getString(R.string.EVALUA_4_M3_SI_1))
            ),
            listOf(
                Child(getString(R.string.EVALUA_4_M4_SI_1)),
                Child(getString(R.string.EVALUA_4_M4_SI_2)),
                Child(getString(R.string.EVALUA_4_VALORACION_GLOBAL_LECTURA)),
                Child(getString(R.string.EVALUA_4_INDICE_GENERAL_LECTURA))
            ),
            listOf(
                Child(getString(R.string.EVALUA_4_M5_SI_1)),
                Child(getString(R.string.EVALUA_4_INDICE_GENERAL_ESCRITURA))
            ),
            listOf(
                Child(getString(R.string.EVALUA_4_M6_SI_1)),
                Child(getString(R.string.EVALUA_4_M6_SI_2)),
                Child(getString(R.string.EVALUA_4_VALORACION_GLOBAL_MATEMATICAS)),
                Child(getString(R.string.EVALUA_4_INDICE_GENERAL_MATEMATICO))
            )
        )

        binding.rv4.apply {
            adapter = HeaderAdapter(
                getString(R.string.routeMapEvalua4),
                headers,
                subItemsList,
                applicationContext,
                this@Evalua4Activity
            )
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(this@Evalua4Activity)
        }
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        if (item.itemId == android.R.id.home) {
            Timber.d(getString(R.string.ACTIVIDAD_CERRADA))
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}